### ISSUE TITLE
Author Naming Inconsistency Issue

### ISSUE THREAD

**Issue by horsburgh** [hydroshare/hydroshare#3428]
See #3427 for the original report of this issue.

There are issues with the current operation of author naming for users with multiple last names. I believe that this only applies for authors who are not existing HydroShare users because users have separate first, middle, and last names in their profile. 

For example, if an author has the name "Rip Van Winkel", the last name is not always consistently displayed, and it depends on how it was entered into the user interface. This likely applies to names added via the API and names aded via the user interface. The following is a demonstration of adding an author that is not a hydroshare user. In the current user interface, the following is the prompt for entering the author's name:

![image](https://user-images.githubusercontent.com/3485902/59113617-9ab4f080-8902-11e9-899c-5b39a70d9116.png)

It is suggested in the text box that the name be entered as "Smith, John Arnold". So, entering "Rip Van Winkel" as "Van Winkel, Rip" results in the following display at the top of the landing page:

![image](https://user-images.githubusercontent.com/3485902/59113713-d0f27000-8902-11e9-9615-314762416276.png)

The authors are inconsistently displayed because my name is first name first, but the other name is last name first. I think they should both be consistent and list the first name first. The the following shows up in the How to Cite section:

![image](https://user-images.githubusercontent.com/3485902/59113773-ee273e80-8902-11e9-8333-41c9e5a74401.png)

This is correctly formatted and the desired result.

If instead I enter the name as "Rip Van Winkel" on the user interface (or I'm assuming the functionality is the same if it gets entered through the API), I get the following at the top of the landing page:

![image](https://user-images.githubusercontent.com/3485902/59113900-37778e00-8903-11e9-9d47-0e95ec2ea168.png)

This is great and looks like what I would expect. However, the following shows up in the How to cite section:

![image](https://user-images.githubusercontent.com/3485902/59113933-49f1c780-8903-11e9-8f9e-fb6f2aab4aac.png)

Part of the last name is not included in the initials where it shouldn't be.

Given that we have a problem of inconsistently entered names via the user interface and the API, I suggest we take another look at separating first, middle, and last name into separate input fields. This is how it is in the HydroShare profile, and it would make the entry of names for authors who are not HydroShare users consistent with what we are asking for in the profile. Then, we could make sure we are always displaying names consistently.

**Comment by dtarb** [comment]
Thanks for working on this.  See also discussion in #855 and #1021 where this was worked on previously.

**Comment by horsburgh** [comment]
Thanks @dtarb for pointing at those prior issues. Based on my reading of those previous issues, the functionality is working as expected, with the exception of the inconsistent listing of author names at the top of the resource landing page:

![image](https://user-images.githubusercontent.com/3485902/59148408-f3998d00-89c5-11e9-9911-895cce7427a0.png)

I suggest that the quick fix here is to fix this rendering issue. When an author's name is entered with the last name first and then a comma and then the first name(s), it should be rendered the same as if the author's name was extracted from an existing user's profile with first name and then last name.

@martinseul - you reported this issue after conversations with the CZO people. They should be able to solve the multiple last name issue by entering the last name as they want it to appear (even with spaces) followed by a comma and then followed by the first name (e.g., "Horsburgh, Jeffery S." or "Van Winkle, Rip").  If we fix this small rendering issue, then I think things are working as they were designed to work.

**Comment by dtarb** [comment]
@horsburgh your suggestion may not fully get us out of the woods.  If there is a user with first name Rip and last name "Van Winkle" who creates a resource, it needs to be rendered into the name field of the Resource author as Van Winkle, Rip, not Rip Van Winkle as it currently is rendered as I just tested.
I created user
![image](https://user-images.githubusercontent.com/3456833/59153991-2030c180-8a25-11e9-846b-774ba2ae699d.png)

Then I created a resource and the citation is stated as 
Winkle, R. V. (2019). test, HydroShare, http://www.hydroshare.org/resource/5758e12575e14981bcfacb7c42619647

**Comment by horsburgh** [comment]
OK - if this is still broken for new users who have multiple last names, I suggest a more comprehensive fix by doing the following:

* [ ] Change the entry of Author names for authors that are not HydroShare users to include first, middle, and last names, consistent with name entry on the HydroShare user profile page.
* [ ] Change the display of Author names at the top of the resource landing page to list authors as "First Middle Last" in all cases.
* [ ] Change the display of Author names in the citation so that it handles multiple last names correctly for all users. Names should appear as "Last, First Initial. Middle Initial." (note the periods after the initials)
* [ ] Change the REST API to allow entry of First, Middle, Last as three separate elements

Then, we have to address all of the resources for which users have now entered Author names as one element. The following rules would still have to apply:

* When a user has entered a name with a comma, anything before the comma is treated as the last name, and anything after the comma is treated as First and Middle name(s)
* When a user has entered a name with multiple elements, but no comma, the string is parsed with the last element being the last name and any other elements being first and middle name(s).

@dtarb - am I missing anything here?

**Comment by dtarb** [comment]
@horsburgh this is a change to the person model within the resource data model so is fairly weighty.  How would it handle things like multiple middle names.  How would it handle suffixes like Jr or the III.  We have discussed some of these before but never fully gotten to closure.

A less weighty lift would be for the copying of names from the profile into the resource name to use the comma approach.  The problem I illustrated above is caused by
Firstname: Rip
Lastname: Van Winkle
being copied into 
Name: Rip Van Winkle
in the resource which parses into Winkle, R. V. in the citation.
If instead it was copied into the resource as Van Winkle, Rip, as we are suggesting users with multipart lastnames should be entered, then the citation would parse into Van Winkle, R.

We may benefit a bit from looking at how citation managers such as Endnote (which is where the comma idea came from) and others handle this.

**Comment by horsburgh** [comment]
Sure - the comma approach can work. It just needs to get consistently implemented and the display issues need to get sorted out.

I've assigned this one to @Maurier, but reading back through past issues, I don't think he did the initial implementation. Should I be assigning this to someone else?

**Comment by dtarb** [comment]
I am happy for you and @Maurier to work it out. If back end changes are needed @sblack-usu or @pkdash could help.

**Comment by sblack-usu** [comment]
I fear we're opening a can of worms that'll bite us in the future if we have rendering logic that expects different formats for names.  It's often best to normalize what's stored in the db, then rendering doesn't need any logic around different types of formats.

So, to follow what's been outline here, I suggest instead of looking for and displaying different name formats on the frontend, we look for different name formats in the db and as the user adds them, normalize the formatting and store all names the same way.

**Comment by Maurier** [comment]
@dtarb, @horsburgh, @sblack-usu Changing the way the data is copied to the resource as @dtarb describes above will indeed address the issue for new entries, but Author data that was copied to the database will remain the same. 

If I change they way new data is copied so that it uses a comma approach, do we have a plan to address existing data?

**Comment by horsburgh** [comment]
@Maurier - correct. Even with the short term fix, we will still have some existing resources for which author data was copied into the database as First Middle Last. The short term fix stops this from happening for more resources. 

The longer term fix is to separate first, middle, and last into three separate database fields. To migrate data for existing resources, the following two rules will be needed:

1. For authors with a comma in their name in the database - anything before the comma would be assumed to be the last name, and anything after the comma would be assumed to be the first name. We won't be able to resolve first versus middle names - both will be copied into the first name field. Users can edit this if they need to, but since multiple first names should be rendered on the page the same as a first + middle name, it shouldn't matter from a display perspective.
2. For authors with no comma in their name in the database - we would assume the last term in the name is the last name. Again, since we can't resolve first and middle names, any names before the last term would get copied into the first name field. Users can edit as needed.

These would have to be addressed using a migration that @sblack-usu would likely write. @sblack-usu and @Maurier - do you guys think these two rules would handle what's in the database now?

**Comment by Maurier** [comment]
@horsburgh I have deployed changes with the short-term fix at https://riverdev.hydroshare.org
Let me know if you see any issues.

**Comment by miguelcleon** [comment]
> 
> 
> Thanks @dtarb for pointing at those prior issues. Based on my reading of those previous issues, the functionality is working as expected, with the exception of the inconsistent listing of author names at the top of the resource landing page:
> 
> ![image](https://user-images.githubusercontent.com/3485902/59148408-f3998d00-89c5-11e9-9911-895cce7427a0.png)
> 
> I suggest that the quick fix here is to fix this rendering issue. When an author's name is entered with the last name first and then a comma and then the first name(s), it should be rendered the same as if the author's name was extracted from an existing user's profile with first name and then last name.
> 
> @martinseul - you reported this issue after conversations with the CZO people. They should be able to solve the multiple last name issue by entering the last name as they want it to appear (even with spaces) followed by a comma and then followed by the first name (e.g., "Horsburgh, Jeffery S." or "Van Winkle, Rip"). If we fix this small rendering issue, then I think things are working as they were designed to work.


@horsburgh I tried this and it worked as described for one of our resources: https://www.hydroshare.org/resource/c646475a6c224f5d838f450ba0947f01/ 

unfortunately when I tried to fix the another one, it broke the resource. I'm not sure what went wrong or how I can get the resource to display again. It just shows an error message now. The broken resource is here: https://www.hydroshare.org/resource/043dccd3fa98474d8cdd15e42231a08d/ 

Breaking the name into three fields, first, middle, and last seems like a good idea.

**Comment by horsburgh** [comment]
@miguelcleon - @Maurier deployed a fix to one of our testing servers and I tested it this morning. It fixes the inconsistency in display of Author names, where the author name was being displayed differently at the top of the resource landing page if you entered the name with last name first and a comma versus if you entered first and then the last name. Also, when adding authors that are HydroShare users, it was copying them into the resource without the comma, so multi-part last names were not appearing correctly in the citation.

The recommended approach for author name entry is Lastname, Firstname Middlename. This is true for both the web user interface and the API. If we can get Mauriel's fix reviewed relatively quickly it should go in the next release. We are still discussing the split into 3 separate fields in the database because it would require a bit of effort and a migration for existing resources. The fix that @Maurier has done so far largely takes care of the main pain points we've been having.

Hopefully @sblack-usu can have a look at your broken resource and see what happened. If @sblack-usu is not the right person, he should know who to pass it off to to get an answer.

**Comment by Maurier** [comment]
@sblack-usu has deployed this fix with actual data at https://dev-hs-1.cuahsi.org for further testing.

**Comment by sblack-usu** [comment]
@horsburgh @miguelcleon - I fixed the resource with a temporary change.  I'm logging an issue to get a code fix so this won't happen again.

We're getting bit by non-ascii characters recently.  We're working on a full long term while doing smaller patches in the mean time.  The resource was broken because one of the authors (Chapela Lara, Marìa) had a non-ascii character in her name (Marìa).  To fix the resource immeidately, I changed the spelling to use an i instead of ì.  We can change it back after the fix is patched in the code.  I'll try to turn it around for the 1.24 release on July 2nd.
